src/components/impagination.js

```diff
@@ -0,0 +1,149 @@
+import { Component } from 'react';
+import PropTypes from 'prop-types';
+import Dataset from 'impagination';
+
+export default class Impagination extends Component {
+  static propTypes = {
+    pageSize: PropTypes.number,
+    loadHorizon: PropTypes.number,
+    readOffset: PropTypes.number,
+    totalPages: PropTypes.number,
+    fetch: PropTypes.func.isRequired,
+    collection: PropTypes.object.isRequired,
+    children: PropTypes.func.isRequired
+  };
+
+  static defaultProps = {
+    pageSize: 25,
+    loadHorizon: 25,
+    readOffset: 0
+  };
+
+  constructor(props) {
+    super(props);
+
+    // track promises for each page to resolve in
+    // `componentWillReceiveProps` later
+    this.promises = {};
+
+    // impagination dataset
+    this.dataset = new Dataset({
+      pageSize: props.pageSize,
+      loadHorizon: props.loadHorizon,
+      stats: { totalPages: props.totalPages },
+      fetch: this.fetch,
+      observe: this.observe
+    });
+
+    // initial state
+    this.state = {
+      datasetState: this.dataset.state
+    };
+  }
+
+  // set the initial read offset of our dataset
+  componentWillMount() {
+    let { readOffset } = this.props;
+    this.dataset.setReadOffset(readOffset);
+  }
+
+  // when we recieve an update, loop over the pages of collections so
+  // we can resolve or reject any pending requests
+  componentWillReceiveProps(nextProps) {
+    let { totalPages, collection } = nextProps;
+
+    // new, zero length collections are likely entirely new
+    let isNewCollection = !collection.length &&
+      collection.isLoading !== this.props.collection.isLoading;
+
+    // we need to reset the dataset
+    if (isNewCollection) {
```

Comment:
Referential equality is not good enough because whenever any part of our data store changes, a new resolver is created, which in turn creates new Models and Collections. So when any request is resolved, that creates a new Collection.

I did throw away the old dataset before, but then I found `reset()` and though it was a good use here. I'll go back to creating a new dataset.